### PR TITLE
fixed #17012: Loading assets synchronously causes rendering to freeze.

### LINKED ISSUE #17012: Loading assets synchronously causes rendering to freeze
Reported at https://forum.cocos.org/t/topic/154102

Using ThreadPool to load assets as which we do for image loading may be an option.

### PR DESCRIPTION
**Pull request by dumganhar**
The main thread will not be freezed now and the performance on android is better than before.

Re: #17012

https://github.com/cocos/cocos-engine-external/pull/455

## Performance

| Devices                       | Loading 200+ json files (size: 90MB) 3 times (Before), unit: milliseconds | Loading 200+ json files (size: 90MB) 3 times (Now), unit: milliseconds |
| ----------------------------- | ---------------------------------------------------- | ------------------------------------------------- |
| SM-J400G (arm32, android 8.0) | 4346<br/>4324<br/>4326                               | 2698<br/>3194<br/>2877                            |
| Pixel 4XL (arm64, android 13) | 971<br/>964<br/>975                                  | 699<br/>662<br/>713                               |
| SM-J111F (arm32, android 5.1) | 5871<br/>5939<br/>5622                               | 3662<br/>3983<br/>3704                            |
| iphone X (arm64, iOS 16.7.2)  | 696<br/>664<br/>647                                  | 696<br/>708<br/>695                               |
| ipad 6 (arm64, iOS 12.2)      | 674<br/>755<br/>717                                  | 736<br/>738<br/>729                               |
| mac m1pro (arm64, macOS 14.5) | 359<br/>424<br/>404                                  | 410<br/>452<br/>408                               |

### Changelog

*

-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

## Greptile Summary

Introduced asynchronous asset loading to prevent rendering freezes, significantly improving performance on Android devices.

- **`native/cocos/bindings/manual/jsb_global.cpp`**: Added async image loading and saving functions, leveraging a thread pool.
- **`native/cocos/bindings/jswrapper/v8/Utils.h`**: Enhanced `ExternalStringResource` with a `Dispose` method for better memory management.
- **`native/cocos/bindings/manual/jsb_cocos_manual.cpp`**: Implemented async asset loading to avoid rendering freezes, focusing on thread pool tasks and JSON parsing.
- **`native/external-config.json`**: Updated to reference `v3.8.4-3` of the external engine-native repository.
- **`cocos/native-binding/index.ts`**: Introduced async file reading functions (`readTextFile`, `readJsonFile`, `readDataFile`).

### CODE REVIEW

File: native/cocos/bindings/manual/jsb_global.cpp [hunk 1]
```diff
@@ -1526,7 +1526,7 @@ SE_BIND_FUNC(JSB_openharmony_postSyncMessage)
 #endif
 
 bool jsb_register_global_variables(se::Object *global) { // NOLINT
-    gThreadPool = LegacyThreadPool::newFixedThreadPool(3);
+    gIOThreadPool = LegacyThreadPool::newFixedThreadPool(5);
```
Comment (dumganhar) [inline native/cocos/bindings/manual/jsb_global.cpp:1558]: There may be lots of json/ccb files to be loaded, so add 2 more threads in pool.

File: native/cocos/bindings/jswrapper/v8/Object.h [hunk 2]
```diff
@@ -152,6 +152,7 @@ class Object final : public RefCounter {
      *  @note The return value (non-null) has to be released manually.
      */
     static Object *createJSONObject(const ccstd::string &jsonStr);
+    static Object *createJSONObject(std::u16string &jsonStr);
```
Comment (minggo) [inline native/cocos/bindings/jswrapper/v8/Object.h:162]: It is better to add a comment saying that, jsonStr is moved in `createJSONObject `.

File: native/cocos/bindings/manual/jsb_global.cpp [hunk 3]
```diff
@@ -1619,6 +1620,6 @@ bool jsb_register_global_variables(se::Object *global) { // NOLINT
         SAFE_DEC_REF(__jsbObj);
         SAFE_DEC_REF(__glObj);
     });
-
+
```
Comment (minggo) [inline native/cocos/bindings/manual/jsb_global.cpp]: extra spaces.

File: native/cocos/bindings/jswrapper/v8/Object.h (hunk 2 above)
Comment (dumganhar) [reply native/cocos/bindings/jswrapper/v8/Object.h:162]: Updated.

File: native/cocos/bindings/manual/jsb_global.cpp (hunk 3 above)
Comment (dumganhar) [reply native/cocos/bindings/manual/jsb_global.cpp]: Fixed.

**Comment by dumganhar** [comment]
@cocos-robot run test cases

**Review by minggo — APPROVED** [review]

**Pull request merged**